vitrina/smart_contracts/views.py

```diff
@@ -0,0 +1,124 @@
+from itertools import groupby
+from typing import Any
+
+from django.contrib import messages
+from django.contrib.auth.mixins import LoginRequiredMixin, PermissionRequiredMixin
+from django.core.handlers.wsgi import WSGIRequest
+from django.forms import modelformset_factory, BaseFormSet
+from django.http import HttpResponseRedirect
+from django.http.response import HttpResponseBase, HttpResponse
+from django.shortcuts import get_object_or_404
+from django.urls import reverse
+from django.utils.translation import gettext_lazy as _
+
+from vitrina.orgs.models import Organization
+from vitrina.orgs.services import has_perm, Action
+from vitrina.projects.models import Project
+from vitrina.smart_contracts import AgreementStatuses
+from vitrina.smart_contracts.forms import (
+    SmartContractForm,
+    SmartContractFormSetHelper,
+)
+from vitrina.smart_contracts.models import Agreement, AgreementScope
+from vitrina.views import FormsetView
+
+
+class AgreementCreateView(
+    LoginRequiredMixin,
+    PermissionRequiredMixin,
+    FormsetView,
+):
+    object: Project
+
+    model = Project
+    template_name = "smart_contracts/agreement_create.html"
+
+    def has_permission(self) -> bool:
+        return (
+            has_perm(self.request.user, Action.UPDATE, self.object)
+            or self.request.user == self.object.user
+        )
+
+    def dispatch(
+        self, request: WSGIRequest, *args: Any, **kwargs: Any
+    ) -> HttpResponseBase:
+        self.object = get_object_or_404(
+            Project.public.all().prefetch_related("datasets"), pk=kwargs["pk"]
+        )
+        if Agreement.objects.filter(project=self.object).exists():
+            messages.error(
+                self.request,
+                _("Šis panaudojimo atvejis jau turi egzistuojančią sutartį."),
+            )
+            return HttpResponseRedirect(
+                reverse("project-datasets", kwargs={"pk": self.kwargs.get("pk")})
+            )
+
+        return super().dispatch(request, *args, **kwargs)
+
+    def get_formset(self) -> BaseFormSet:
+        formset_kwargs = self.get_formset_kwargs()
+        datasets = self.object.datasets.all().order_by("organization_id")
```

Comment:
I might see a catch here: do we get the `deleted=True` datasets here as well? Please check the other queries as well and their models, if they're hiding one of those fields.